### PR TITLE
audio: codec_adapter: Added DTS codec

### PR DESCRIPTION
**Pull request by markbartonxperi**
Added DTS Codec using the codec_adapter API

Signed-off-by: Mark Barton <mark.barton@xperi.com>

### CODE REVIEW

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by dbaluta** [comment]
There is no need for separate Pull Requests. Just separate patches. So this https://github.com/thesofproject/sof/pull/4894 it is not needed. Just add the patch here.

File: src/audio/codec_adapter/codec/dts.c [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+	DtsSofInterfaceVersionInfo interface_version;
+	DtsSofInterfaceVersionInfo sdk_version;
+
+	comp_dbg(dev, "dts_codec_init() start");
+
+	dts_result = dtsSofInterfaceInit((DtsSofInterfaceInst **)&(codec->private),
+		dts_effect_allocate_codec_memory, dev);
+	ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+
+	if (!ret) {
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:128]: nit, flip this with case below and return on error to reduce indent

File: src/audio/codec_adapter/codec/dts.c [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	DtsSofInterfaceResult dts_result;
+
+	comp_dbg(dev, "dts_codec_apply_config() start");
+
+	config = &codec->r_cfg;
+
+	/* Check that config->data isn't invalid and has size greater than 0 */
+	config_header_size = sizeof(config->size) + sizeof(config->avail);
+	if (config->size < config_header_size) {
+		comp_dbg(dev, "dts_codec_apply_config() config->data is invalid");
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:258]: this sounds like it should be an error

File: src/audio/codec_adapter/codec/dts.c [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+		return 0;
+	} else if (config->size == config_header_size) {
+		comp_dbg(dev, "dts_codec_apply_config() size of config->data is 0");
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:261]: this sounds like -EINVAL, we should be returning errors as they get propagated to userspace so we don't fail silently

File: src/audio/codec_adapter/codec/dts.c [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+		return 0;
+	}
+
+	/* Calculate size of config->data */
+	config_data_size = config->size - config_header_size;
+
+	/* Check that config->data is not greater than the max expected for DTS data */
+	if (config_data_size > MAX_EXPECTED_DTS_CONFIG_DATA_SIZE) {
+		comp_dbg(dev,
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:270]: this is an error path, not a debug branch

File: src/audio/codec_adapter/codec/dts.c [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+	/* Allow for multiple codec_params to be packed into the data pointed to by config
+	 */
+	for (i = 0; i < config_data_size; param_number++) {
+		param = (struct codec_param *)((char *)config->data + i);
+		param_header_size = sizeof(param->id) + sizeof(param->size);
+
+		/* If param->size is less than param_header_size, then this param is not valid */
+		if (param->size < param_header_size) {
+			comp_dbg(dev, "dts_codec_apply_config() param is invalid");
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:283]: error path

File: src/audio/codec_adapter/codec/dts.c [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+				}
+			}
+		}
+
+		/* Advance to the next codec_param */
+		i += param->size;
+	}
+
+	if (ret)
+		comp_err(dev, "dts_codec_apply_config() failed %d", ret);
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:310]: what is this catching? All functions that can fail have logs already

File: src/audio/codec_adapter/codec/dts.c [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+			if (param_data_size) {
+				dts_result = dtsSofInterfaceApplyConfig(codec->private, param->id,
+					param->data, param_data_size);
+				ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+				if (ret) {
+					comp_err(dev,
+						"dts_codec_apply_config() dtsSofInterfaceApplyConfig failed %d",
+						dts_result);
+					break;
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:300]: Can we recover from this? If not we should exit not continue (prevent future modifications adding code on an error path by accident)

**Review by cujomalainey — CHANGES_REQUESTED** [review]

File: src/audio/codec_adapter/codec/dts.c [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+DECLARE_SOF_RT_UUID("dts_codec", dts_uuid, 0xd95fc34f, 0x370f, 0x4ac7,
+			0xbc, 0x86, 0xbf, 0xdc, 0x5b, 0xe2, 0x41, 0xe6);
+DECLARE_TR_CTX(dts_tr, SOF_UUID(dts_uuid), LOG_LEVEL_INFO);
+
+#define MAX_EXPECTED_DTS_CONFIG_DATA_SIZE 8192
+
+static void *dts_effect_allocate_codec_memory(void *dev_void, unsigned int length,
+	unsigned int alignment)
+{
+	struct comp_dev *dev = (struct comp_dev *)dev_void;
```
Comment (cujomalainey) [inline src/audio/codec_adapter/codec/dts.c:22]: i don't think you need this cast, but i might be wrong

File: tools/topology/topology1/sof/pipe-eq-iir-codec-adapter-playback.m4 [hunk 9]
```diff
@@ -0,0 +1,199 @@
+# Low Latency Passthrough Pipeline and PCM
+#
```
[the remark this answers is not in the document]
Comment (markbartonxperi) [reply tools/topology/topology1/sof/pipe-eq-iir-codec-adapter-playback.m4:2]: Changed

File: src/include/sof/audio/dts_sof_interface/DtsSofInterface.h [hunk 10]
```diff
@@ -0,0 +1,110 @@
+/*======================================================================*
+ DTS, Inc.
+ 5220 Las Virgenes Road
+ Calabasas, CA 91302  USA
+
+ CONFIDENTIAL: CONTAINS CONFIDENTIAL PROPRIETARY INFORMATION OWNED BY DTS, INC. AND/OR ITS AFFILIATES ("DTS"), INCLUDING
+BUT NOT LIMITED TO TRADE SECRETS, KNOW-HOW, TECHNICAL AND BUSINESS INFORMATION. USE, DISCLOSURE OR DISTRIBUTION OF THE
```
[the remark this answers is not in the document]
Comment (markbartonxperi) [reply src/include/sof/audio/dts_sof_interface/DtsSofInterface.h]: Changed to `comp_err` and return `-EINVAL`

File: src/audio/codec_adapter/CMakeLists.txt [hunk 11]
```diff
@@ -57,3 +57,10 @@ sof_add_static_library(MaxxChrome ${CMAKE_CURRENT_LIST_DIR}/lib/release/libMaxxC
 # folder with Waves API must be among include directories
 target_include_directories(sof PUBLIC ${CMAKE_CURRENT_LIST_DIR}/../../include/sof/audio/)
 endif()
```
[the remark this answers is not in the document]
Comment (markbartonxperi) [reply src/audio/codec_adapter/CMakeLists.txt:59]: Changed to `comp_err` and return `-EINVAL`

File: src/audio/codec_adapter/codec/dts.c [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+	if (ret == 0) {
+		dts_result = dtsSofInterfacePrepare(
+			(DtsSofInterfaceInst *)codec->private,
+			&buffer_configuration,
+			&codec->cpd.in_buff,
+			&codec->cpd.in_buff_size,
+			&codec->cpd.out_buff,
+			&codec->cpd.out_buff_size);
+		ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+		if (ret != 0)
```
[the remark this answers is not in the document]
Comment (markbartonxperi) [reply src/audio/codec_adapter/codec/dts.c]: Changed to `comp_err` and return `-EINVAL`

File: src/audio/codec_adapter/codec/dts.c [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+	} else if (config->size == config_header_size) {
+		comp_dbg(dev, "dts_codec_apply_config() size of config->data is 0");
+		return 0;
+	}
+
+	/* Calculate size of config->data */
+	config_data_size = config->size - config_header_size;
+
+	/* Check that config->data is not greater than the max expected */
+	if (config_data_size > MAX_EXPECTED_CONFIG_DATA_SIZE) {
```
[the remark this answers is not in the document]
Comment (markbartonxperi) [reply src/audio/codec_adapter/codec/dts.c]: Changed to `comp_err` and return `-EINVAL`

File: src/audio/codec_adapter/codec/dts.c (hunk 1 above)
Comment (markbartonxperi) [reply src/audio/codec_adapter/codec/dts.c]: Good point. Now removed.

File: src/audio/codec_adapter/codec/dts.c (hunk 3 above)
Comment (markbartonxperi) [reply src/audio/codec_adapter/codec/dts.c]: Removed the cast

File: src/audio/codec_adapter/codec/dts.c (hunk 2 above)
Comment (markbartonxperi) [reply src/audio/codec_adapter/codec/dts.c]: Whilst we possibly could recover, I think it's safest here to return `-EINVAL` at the first sign of an individual config blob failing to be applied in our codec. Changed code to early out with error.

**Review by cujomalainey — APPROVED** [review]

**Comment by cujomalainey** [comment]
3 maintainers have signed off now, merging

**Pull request merged**